docs/Toy/toy/parser.py

```diff
@@ -0,0 +1,437 @@
+from pathlib import Path
+from typing import List, NoReturn, TypeVar, cast
+
+from .lexer import (OperatorToken, Token, tokenize, NumberToken, EOFToken,
+                    IdentifierToken)
+from .toy_ast import (BinaryExprAST, ExprAST, NumberExprAST, FunctionAST,
+                      ModuleAST, LiteralExprAST, VarDeclExprAST, VarType,
+                      VariableExprAST, ReturnExprAST, PrintExprAST,
+                      CallExprAST, PrototypeAST)
+
+
+class ParseError(Exception):
+
+    def __init__(self,
+                 token: Token,
+                 expected: str | type[Token],
+                 context: str = '',
+                 line: str = ''):
+        loc = token.loc
+
+        message = f'Parse error ({loc.line}, {loc.col}): expected '
+        if isinstance(expected, str):
+            message += expected
+        else:
+            message += expected.name()
+
+        if len(context):
+            message += ' ' + context
+
+        message += f" but has Token '{token.text}'\n"
+
+        if len(line):
+            message += line + '\n'
+
+        super().__init__(message)
+
+    pass
+
+
+TokenT = TypeVar('TokenT', bound=Token)
+
+
+class Parser:
+    file: Path
+    program: str
+    tokens: List[Token]
+    pos: int
+
+    def __init__(self, file: Path, program: str):
+        self.file = file
+        self.program = program
+        self.tokens = tokenize(file, program)
+        self.pos = 0
+
+    def getToken(self):
+        '''Returns current token in parser'''
+        return self.tokens[self.pos]
+
+    def getTokenPrecedence(self) -> int:
+        '''Returns precedence if the current token is a binary operation, -1 otherwise'''
+        PRECEDENCE = {
+            '-': 20,
+            '+': 20,
+            '*': 40,
+        }
+        op = self.getToken().text
+
+        try:
+            return PRECEDENCE[op]
+        except KeyError:
+            return -1
+
+    def peek(self, pattern: str | type[Token] | None = None) -> None:
+        '''Verifies that the current token fits the pattern, raises ParseError otherwise'''
+        token = self.getToken()
+        tokenType, text = (None,
+                           pattern) if isinstance(pattern, str) else (pattern,
+                                                                      None)
+        if tokenType is not None:
+            if type(token) is not tokenType:
+                self.parseError(tokenType)
+
+        if text is not None:
+            if token.text != text:
+                self.parseError(f"'{text}'")
+
+    def check(self, pattern: str | type[Token] | None = None) -> bool:
+        '''Verifies that the current token fits the pattern, raises ParseError otherwise'''
+        try:
+            self.peek(pattern)
+            return True
+        except ParseError:
+            return False
+
+    def pop(self) -> Token:
+        self.peek()
+        self.pos += 1
```

Comment:
why do not we just do return self.tokens[self.pos] ? Am I missing something?

Reply:
Because I need to change pos before returning it. I could do something like `res = self.tokens[self.pos]; self.pos += 1; return res` but I prefer how I wrote it in the code